File: leetcode/3/main.py
# 3. Longest Substring Without Repeating Characters


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left, longest = 0, 0
        chars = set()

        for r in range(len(s)):
            # if new index in set, keep removing
            while s[r] in chars:
                chars.remove(s[left])
                left += 1

            # then add new index
            chars.add(s[r])
            # update longest
            longest = max(longest, r - left + 1)

        return longest

File: leetcode/3/main.go
// 3. Longest Substring Without Repeating Characters

package main

// first attempt
func lengthOfLongestSubstring(s string) int {
	l, r, long := 0, 0, 0
	for r < len(s) {
		curr := s[l : r+1]
		if unique(curr) && len(curr) > long {
			long = len(curr)
		} else {
			l++
		}
		r++
	}

	return long
}

func unique(arr string) bool {
	m := make(map[rune]bool)
	for _, i := range arr {
		if _, ok := m[i]; ok {
			return false
		}

		m[i] = true
	}
	return true
}

// use a set like neetcode
func lengthOfLongestSubstringSet(s string) int {
	l, long := 0, 0
	m := make(map[byte]bool)
	for r := range s {
		_, in := m[s[r]]
		for in {
			delete(m, s[l])
			l++
			in = m[s[r]]
		}
		m[s[r]] = true
		long = max(long, r-l+1)
	}

	return long
}
